include/mbedtls/mps/writer.h

```diff
@@ -0,0 +1,825 @@
+/**
+ * \file writer.h
+ *
+ * \brief This file defines writer objects, which together with
+ *        their sibling reader objects form the basis for the communication
+ *        between the various layers of the Mbed TLS messaging stack,
+ *        as well as the communication between the messaging stack and
+ *        the (D)TLS handshake protocol implementation.
+ *
+ * Writers provide a means of communication between an
+ * entity (the 'provider' in the following) providing buffers
+ * to which outgoing data can be written, and an entity
+ * (the 'consumer' in the following) consuming them by writing
+ * the actual data into it.
+ * Both the size of the data buffers the provider prepares
+ * and the size of chunks in which the consumer writes the
+ * data are variable and may be different.
+ * It is the writer's responsibility to do the
+ * necessary copying and pointer arithmetic.
+ *
+ * The basic flow of operation is that the provider feeds
+ * an outgoing data buffer to the writer, transferring it from
+ * 'providing' to 'consuming' mode. The consumer subsequently
+ * fetches parts of the buffer and writes data to them.
+ * Once that's done, the provider revokes the writer's access
+ * to the outgoing data buffer, putting the writer back to
+ * providing mode; the provider may then continue processing
+ * (e.g. dispatching) the data provided in the outgoing data buffer.
+ * After that, the provider feeds another outgoing data buffer
+ * to the writer and the cycle starts again.
```

Comment:
@k-stachowiak I have rewritten the introduction today, could you have a look and see if you find it clearer? I'm also not sure about the 'producer' / 'consumer' nomenclature - I feel it could be made clearer. Maybe 'dispatcher' and 'writer' is a better fit?